### DISCUSSION TITLE
ESP8266 Setup Parameters

### DISCUSSION THREAD

**Discussion by McJohnSrl** [JK-Rolling/DuinoI2C_ESP#5] [Ideas]
Hi @JK-Rolling 
I would like to suggest if it is possible to insert two fields in the initial setup, one to set the LED pin and the other to set I2C_SCL (100 or 400) KHz.
I think this reduces the production of many firmwares with different variants.
Thank you :)

![immagine](https://github.com/user-attachments/assets/d5a27fa3-a302-4625-a89d-601906740de3)

**Comment by JK-Rolling** [comment]
yes, I agree. I was indecisive whether to add these fields into the setup page. my worries was the user may not know what are these fields for and end up breaking i2c.

let me think about it. Thanks for the suggestion!

**Reply by McJohnSrl** [reply]
I understand your concern, you could propose some basic default values ​​so that the undecided user can't make a mistake.

**Reply by JK-Rolling** [reply]
there are a few other user parameter that I can add. for example
* testbench mode
* job diff level - (AVR, ARM)
* group workers together in wallet

but all these makes the config page became too long, bigger .bin size, 
and it'd means there are more ways to break the firmware operation

**Reply by McJohnSrl** [reply]
you could split the setup page into two pages, the first one named Basic Setup with the current parameters, and the second one named Advanced Setup with the new fields.
It is understood that all advanced fields must have the field with the default basic value.

**Reply by JK-Rolling** [reply]
great idea. v0.24 already have collapsible fields for network setting. maybe I can add another collapsible for advanced settings

**Comment by JK-Rolling** [comment]
Interim update.
screen shot 1
![image](https://github.com/user-attachments/assets/38884ecd-fc50-4b70-93e2-110c629d3ad4)

screen shot 2
![image](https://github.com/user-attachments/assets/1861be8d-9c44-496c-a102-8cd0484d78c0)

**Reply by McJohnSrl** [reply]
it's perfect thanks :)

**Reply by McJohnSrl** [reply]
please explain what "Group Wolkers" is for

**Reply by JK-Rolling** [reply]
![image](https://github.com/user-attachments/assets/1cf7a01b-8794-4575-94ed-bffb46c35249)

**Comment by McJohnSrl** [comment]
Regarding the problem that ESP8266 has difficulty managing more than 10 workers because it has little memory available, I thought that is possible use 2 ESP8266 masters on the same bus and divide the workload of the workers.
For this is need two new parameters in the advanced setup: "I2C Address From" and "I2C Address To" in this way I can set the I2C addresses from 0 to 9 for the first master and from 10 to 19 for the second master.
What do you think of this idea, do you like it ?

**Reply by JK-Rolling** [reply]
Technically I2C specification do allow multi-master where collision detection and arbitration is involved, but it has challenges of it's own like adding processing overhead, shifting timing, added clock stretching that slows down SCL. 
It'll be easier to dedicate one master per I2C bus.

**Reply by McJohnSrl** [reply]
So, using only one master,  is possible get better results using Attiny 85 workers which have a 400K I2C clock instead of the 100K of ATMEGA328.
What do you think ?

**Reply by JK-Rolling** [reply]
yes. theoretically it should get better results. 
some experiment data contributed by the community. Summarizing them in the table below
|data|Host|Protocol|Worker|shares|Notes|
|:-:|:-:|:-:|:-:|:-:|:-:|
|1|ESP8266|I2C 100K ↔️ 100K|Attiny85 (4x)|235800 (24hrs)|`Base`|
|2|ESP8266|I2C 100K ↔️ 400K|Attiny85 (4x)|256100 (24hrs)|8.6% better than `Base`|
|3|ESP8266|I2C 400K ↔️ 400K|Attiny85 (4x)|?|Attiny lost ~50H/s, lower than `Base`|
|4|ESP8266|I2C 100K ↔️ 100K|ATM328 (6x)|15648 (1hr)|`Base2`|
|5|ESP8266|I2C 100K ↔️ 400K|ATM328 (6x)|16340 (1hr)|4.3% better than `Base2`|
|6|ESP8266|I2C 400K ↔️ 400K|ATM328 (6x)|-|Failed. Suspecting cloned ESP|
|7|?|USB|? (7x)|460680 (24hrs)|`Base3`|
|8|ESP8266|I2C ? ↔️ ?|? (7x)|458439 (24hrs)|0.5% lower than `Base3`|

**Reply by McJohnSrl** [reply]
what do you mean by (Nx) maybe it's the number of workers used ?

**Reply by JK-Rolling** [reply]
that's right. 4x means 4 workers

**Comment by JK-Rolling** [comment]
enhancement implemented in v0.25. Thanks McJohn for the idea!
